src/datasets/packaged_modules/spark/spark.py

```diff
@@ -29,20 +31,69 @@ class SparkConfig(datasets.BuilderConfig):
     features: Optional[datasets.Features] = None
 
 
+def _generate_iterable_examples(
+    df: "pyspark.sql.DataFrame",
+    partition_order: List[int] = None,
+):
+    import pyspark
+
+    def generate_fn():
+        df_with_partition_id = df.select("*", pyspark.sql.functions.spark_partition_id().alias("part_id"))
+        for partition_id in partition_order:
+            partition_df = df_with_partition_id.select("*").where(f"part_id = {partition_id}").drop("part_id")
+            rows = partition_df.collect()
+            row_id = 0
+            for row in rows:
+                yield f"{partition_id}_{row_id}", row.asDict()
+                row_id += 1
+
+    return generate_fn
+
+
+class SparkExamplesIterable(_BaseExamplesIterable):
+    def __init__(
+        self,
+        df: "pyspark.sql.DataFrame",
+        partition_order=None,
+    ):
+        self.df = df
+        self.generate_examples_fn = _generate_iterable_examples(
+            self.df,
+            partition_order or range(self.df.rdd.getNumPartitions()),
+        )
+
+    def __iter__(self):
+        yield from self.generate_examples_fn()
+
+    def shuffle_data_sources(self, generator: np.random.Generator) -> "SparkExamplesIterable":
+        partition_ids = range(self.df.rdd.getNumPartitions())
+        generator.shuffle(partition_ids)
+        return SparkExamplesIterable(self.df, partition_order=partition_ids)
+
+    def shard_data_sources(self, shard_indices: List[int]) -> "SparkExamplesIterable":
+        return SparkExamplesIterable(self.df, partition_order=shard_indices)
+
+    @property
+    def n_shards(self) -> int:
+        return self.df.rdd.getNumPartitions()
+
+
 class Spark(datasets.DatasetBuilder):
     BUILDER_CONFIG_CLASS = SparkConfig
 
     def __init__(
         self,
         df: "pyspark.sql.DataFrame",
         cache_dir: str = None,
+        streaming: bool = True,
         **config_kwargs,
     ):
         import pyspark
 
         self._spark = pyspark.sql.SparkSession.builder.getOrCreate()
         self.df = df
-        self._validate_cache_dir(cache_dir)
+        if not streaming:
+            self._validate_cache_dir(cache_dir)
```

Comment:
I think you can move this check to `_prepare_split`, this way there's no need to add a new argument `streaming=`

Reply:
Done